File: SearchNShop_and/app/src/main/java/co/esclub/searchnshop/model/SearchItem.java
package co.esclub.searchnshop.model;

import android.os.Parcelable;
import android.util.Log;

import java.util.TreeMap;

import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.Ignore;
import io.realm.annotations.PrimaryKey;

/**
 * Created by tae.kim on 06/06/2017.
 */

public class SearchItem extends RealmObject {
    public static final String KEY_ID = "id";

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getKeyWord() {
        return keyWord;
    }

    public void setKeyWord(String keyWord) {
        this.keyWord = keyWord;
    }

    public String getMallName() {
        return mallName;
    }

    public void setMallName(String mallName) {
        this.mallName = mallName;
    }

    public RealmList<ShopItem> getItems() {
        return items;
    }

    public void setLastSearchTime(long lastSearchTime) {
        this.lastSearchTime = lastSearchTime;
    }

    public long getLastSearchTime() {
        return lastSearchTime;
    }

    public boolean getIsSuccess() {
        return isSuccess;
    }

    public void setIsSuccess(boolean isSuccess) {
        this.isSuccess = isSuccess;
    }

    @PrimaryKey
    private String id;
    private String keyWord;
    private String mallName;
    private long lastSearchTime = 0;

    @Ignore
    private boolean isSuccess = false;

    public TreeMap<Integer, ShopItem> getItemTreeMap() {
        return itemTreeMap;
    }

    public void setItemTreeMap(
            TreeMap<Integer, ShopItem> itemTreeMap) {
        this.itemTreeMap = itemTreeMap;
    }

    @Ignore
    TreeMap<Integer, ShopItem> itemTreeMap = new TreeMap<Integer, ShopItem>();

    RealmList<ShopItem> items = new RealmList<>();

    public SearchItem(String keyWord, String mallName) {
        this.keyWord = keyWord;
        this.mallName = mallName;
        this.id = makeID(keyWord, mallName);
    }

    public SearchItem() {

    }

    public void addItem(ShopItem item) {
        items.add(item);
    }

    public void clearItem() {
        items.clear();
    }

    public static String makeID(String keyWord, String mallName) {
        return keyWord + "_" + mallName;
    }

    public void dump() {
        Log.d("###", "KEY_WORD: " + keyWord);
        Log.d("###", "MALL_NAME: " + mallName);
        if (items.size() > 0) {
            for (ShopItem item : items) {
                Log.d("###", "====");
                Log.d("###", "TITLE: " + item.getTitle());
                Log.d("###", "====");
            }
        } else {
            Log.d("###", "Item size 0");
        }
    }
}

File: SearchNShop_and/app/src/main/java/co/esclub/searchnshop/model/item/SearchItem.kt
package co.esclub.searchnshop.model.item

import co.esclub.searchnshop.util.LogCat
import io.realm.DynamicRealmObject
import java.util.TreeMap

import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.Ignore
import io.realm.annotations.PrimaryKey

/**
 * Created by tae.kim on 06/06/2017.
 */

open class SearchItem : RealmObject, Item {
    @Ignore
    val TAG = SearchItem::class.java.simpleName

    @PrimaryKey
    open var id: String? = null
    open var keyWord: String? = null
    open var mallName: String? = null
    open var lastSearchTime: Long = 0

    @Ignore
    var isSuccess = false

    @Ignore
    var itemTreeMap = TreeMap<Int, ShopItem>()

    open var items = RealmList<ShopItem>()
        internal set

    constructor(keyWord: String?, mallName: String?) {
        this.keyWord = keyWord
        this.mallName = mallName
        this.id = makeID(keyWord, mallName)
    }

    constructor() {
    }

    fun dump() {
        LogCat.d(TAG, "KEY_WORD: " + keyWord!!)
        LogCat.d(TAG, "MALL_NAME: " + mallName!!)
        LogCat.d(TAG, "LAST_UPDATE_TIME" + lastSearchTime)
        if (items.size > 0) {
            for (item in items) {
                LogCat.d(TAG, "====")
                LogCat.d(TAG, "TITLE: " + item.title)
                LogCat.d(TAG, "====")
            }
        } else {
            LogCat.d(TAG, "Item size 0")
        }
    }

    companion object {
        val KEY_ID = "id"

        fun makeID(keyWord: String?, mallName: String?): String {
            return keyWord + "_" + mallName
        }
    }

    constructor(item: DynamicRealmObject) {
        this.keyWord = item.getString("keyWord")
        this.mallName = item.getString("mallName")
        this.lastSearchTime = item.getLong("lastSearchTime")
        for (shopItem in item.getList("items")) {
            shopItem?.let {
                items.add(ShopItem(shopItem))
            }
        }
    }

    fun removeSpacing() {
        this.keyWord = keyWord?.trim()?.replace(" ", "")
        this.id = makeID(keyWord, mallName)
    }

    fun getTransItems(): RealmList<DynamicRealmObject>? {
        val ret  = RealmList<DynamicRealmObject>()
        for(item in items) {
            ret.add(DynamicRealmObject(item))
        }
        return ret
    }
}
